### PR TITLE
Replace `boost::filter_iterator` with explicit iterator implementation

### PR DESCRIPTION
**Pull request by nvmkuruc**
### Description of Change(s)
Converts the `_Predicate` private class into an explicit `_FilterIterator` bidirectional iterator implementation.

### Fixes Issue(s)
- #2311 

- [ ] I have verified that all unit tests pass with the proposed changes

- [x] I have submitted a signed Contributor License Agreement

### CODE REVIEW

**Comment by tallytalwar** [comment]
Filed as internal issue #USD-8056

File: pxr/usd/sdf/childrenView.h [hunk 1]
```diff
@@ -76,27 +75,75 @@ template <typename _Owner, typename _InnerIterator, typename _DummyPredicate>
...
         {
             return _owner->GetPredicate()(
                 _Owner::Adapter::Convert(x));
         }
 
-    private:
-        const _Owner* _owner;
+        void filter()
+        {
+            while (_underlyingIterator != _end &&
```
Comment (crevilla) [inline pxr/usd/sdf/childrenView.h:141]: I'm pretty sure you have this inverted. According to the boost filter_iterator documentation, the iterator retains the element if the predicate returns true and skips the element if the predicate is false. Here, you're skipping when the predicate is true.

File: pxr/usd/sdf/childrenView.h [hunk 2]
```diff
@@ -76,27 +75,75 @@ template <typename _Owner, typename _InnerIterator, typename _DummyPredicate>
...
+            return _underlyingIterator != other._underlyingIterator;
+        }
 
-        _Predicate() : _owner(NULL) { }
-        _Predicate(const _Owner* owner) : _owner(owner) { }
+        const _InnerIterator& GetBase() const { return _underlyingIterator; }
 
-        bool operator()(const value_type& x) const
+    private:
+        // Skip any iterators that don't satisfy the predicate
```
Comment (crevilla) [inline pxr/usd/sdf/childrenView.h:132]: These `shouldFilter` and `filter` functions aren't here for API conformance right? If I'm right, they should follow our coding conventions for private methods, i.e. capitalized and starting with an underscore

File: pxr/usd/sdf/childrenView.h [hunk 3]
```diff
@@ -76,27 +75,75 @@ template <typename _Owner, typename _InnerIterator, typename _DummyPredicate>
...
+    class _FilterIterator {
     public:
-        typedef typename _Owner::value_type value_type;
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = typename _InnerIterator::value_type;
+        using reference = typename _InnerIterator::reference;
+        using pointer = typename _InnerIterator::pointer;
+        using difference_type = typename _InnerIterator::difference_type;
+
+        _FilterIterator() = default;
```
Comment (crevilla) [inline pxr/usd/sdf/childrenView.h:88]: The body of all three of these should be on a new line like you did for all the rest of the functions.

File: pxr/usd/sdf/childrenView.h (hunk 1 above)
Comment (nvmkuruc) [reply pxr/usd/sdf/childrenView.h:141]: That was it.  All tests are now passing.

**Pull request merged**